File: 513. Find Bottom Left Tree Value.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
let findBottomLeftValue = function(root) {
    let output = 0
    let queue = []
    
    if (root == null) {
        return 0
    }
    
    queue.push(root)
    
    while (queue.length > 0) {
        let level = []
        let x = 0
        
        while (queue.length > 0) {
            let node = queue.shift()
            x = node.val
            
            if(node.right) {
                level.push(node.right)
            }
            if (node.left) {
                level.push(node.left)
            }
        }
        output = x
        queue = level
    }

    return output
};

File: 513. Find Bottom Left Tree Value.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def findBottomLeftValue(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def bfs(root):
            q1 = []
            q1.append(root)
            
            while q1:
                q2 = []
                
                while q1:
                    node = q1.pop(0)
                    output = node.val
                    
                    if node.right:
                        q2.append(node.right)
                    if node.left:
                        q2.append(node.left)
                q1 = q2
                
                if not q1:
                    return output    
        
        return bfs(root)
